### ISSUE TITLE
MA0050 only works with "== null", and not "is null"

### ISSUE THREAD

**Issue by perkops** [meziantou/Meziantou.Analyzer#351]
**Context**
Rule identifier: _MA0050_
Target Framework: _.NET 6.0_
Language version: _10.0_

**Describe the bug**

```csharp
    public static IEnumerable<byte[]> Split(
        this IEnumerable<byte> source,
        byte splitByte)
    {
        if (source == null)
        {
            throw new ArgumentNullException(nameof(source));
        }
    }
```

The above code reports an error when using

```c#
if (source is null)
```

The expectation would be that the "is null" is valid code and MA0050 does not flag this as an error.

**Comment by meziantou** [comment]
I can't repro the bug. Could you provide the full repro code? The provided code doesn't compile as the method doesn't return anything.

**Comment by perkops** [comment]
Sorry, today its also hard to re-produce, but it failed last Friday with 'is' and when we changed to '==', the error went away.

The way to reproduce, would be;

1. Clone our repository: https://github.com/atc-net/atc/
2. Change line number (515) in .editorconfig https://github.com/atc-net/atc/blob/main/.editorconfig
    dotnet_diagnostic.MA0050.severity = none    <- to error
3. See that https://github.com/atc-net/atc/blob/main/src/Atc/Extensions/BaseTypes/ByteExtensions.cs wont build

Today it also complains when using '==', so no idea why it worked last Friday.

**Comment by meziantou** [comment]
I don't think there is a bug. The method `ByteExtension.Split` is reported by MA0050 because it doesn't validate its arguments until method is enumerated.

````
IEnumerable<byte> source = null;
var enumerable = source.Split(0); // no exception
enumerable.First(); // ArgumentNullException
````

The suggested fix creates a local function, so the arguments are validated when calling the method.

````
    public static IEnumerable<byte[]> Split(
        this IEnumerable<byte> source,
        byte splitByte)
    {
        if (source is null)
            throw new ArgumentNullException(nameof(source));

        return Split();

        IEnumerable<byte[]> Split()
        {
            var current = new List<byte>();
            // ...
        }
    }
````

**Comment by perkops** [comment]
Thank you very much for the replies and investigation - Implemented fix as suggested by you.
